File: FinLib.NET/FinLib/Options/OptionContract.cs
using System;
using FinLib.Extensions;
using FinLib.Price;

namespace FinLib.Options;

public enum OptionType
{
    Call,
    Put,
}

public enum OptionStyle
{
    European,
    American,
}

public class OptionContract: IDisposable, IProfit<double>
{
    private readonly unsafe OptionContract_native* _handle;
    internal unsafe OptionContract_native* GetPtr() => _handle;

    public OptionContract(OptionType optionType, OptionStyle optionStyle, Side side, double strike, double premium)
    {
        unsafe
        {
            _handle = NativeMethods.option_contract_from(optionType.MapType(), optionStyle.MapStyle(), side.MapSide(), strike, premium);
        }
    }

    public double Payoff(double underlying)
    {
        unsafe
        {
            return NativeMethods.option_contract_payoff(_handle, underlying);
        }
    }

    public double Profit(double underlying)
    {
        unsafe
        {
            return NativeMethods.option_contract_profit(_handle, underlying);
        }
    }

    public bool WillBeExercised(double underlying)
    {
        unsafe
        {
            return NativeMethods.option_contract_will_be_exercised(_handle, underlying);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.option_contract_destroy(_handle);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~OptionContract()
    {
        ReleaseUnmanagedResources();
    }
}

File: finlib-cpp/src/finlib/options/option_contract.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "option_contract.hpp"

namespace finlib {
    OptionContract::OptionContract(finlibrs::OptionType option_type,
                                   finlibrs::OptionStyle option_style,
                                   finlibrs::Side side,
                                   double strike,
                                   double premium) {
        handle = finlibrs::option_contract_from(option_type, option_style, side, strike, premium);
    }

    double OptionContract::payoff(double underlying) {
        return finlibrs::option_contract_payoff(handle, underlying);
    }

    double OptionContract::profit(double underlying) {
        return finlibrs::option_contract_profit(handle, underlying);
    }

    bool OptionContract::will_be_exercised(double underlying) {
        return finlibrs::option_contract_will_be_exercised(handle, underlying);
    }

    OptionContract::~OptionContract() {
        finlibrs::option_contract_destroy(handle);
    }
}
